### PR TITLE
[ignore] Add logic for retrieval and reading meta files into datastore struct (DCNE-343)

### PR DESCRIPTION
**Pull request by akinross**
DCNE-343

### CODE REVIEW

File: gen/utils/data/class.go [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+	ClassNameShort string
+	// List of all child classes which are included inside the resource.
+	// When looping over maps in golang the order of the returned elements is random, thus list is used for order consistency.
+	Children []Child
+	// List of all possible parent classes.
+	ContainedBy []string
+	// Documentation specific information for the class.
+	Documentation ClassDocumentation
+	// List of all identifying properties of the class.
+	// These are properties that are part if the relative name (RN) format, ex 'tn-{name}' where name is the identifying property.
```
Comment (lhercot) [inline gen/utils/data/class.go:32]:
```suggestion
	// These are properties that are part of the relative name (RN) format, ex 'tn-{name}' where name is the identifying property.
```

File: gen/utils/data/class.go [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+	IdentifiedBy []string
+	// Indicates that the class is migrated from previous version of the provider.
+	// This is used to determine if legacy attributes to be exposed in the resource.
```
Comment (lhercot) [inline gen/utils/data/class.go:35]:
```suggestion
	// This is used to determine if legacy attributes have to be exposed in the resource.
```

File: gen/utils/data/class.go [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+	// This is used to determine if the plan modification logic.
```
Comment (lhercot) [inline gen/utils/data/class.go:36]:
```suggestion
	// This is used to determine if the plan modification logic.
```
Is this needed here? Not sure why 2 "this is used"

File: gen/utils/data/class.go [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+	Properties map[string]*Property
+	// Each property is categorised in all, required, optional, or read-only.
+	// When looping over maps in golang the order of the returned elements is random, thus lists are used for order consistency.
+	PropertiesAll      []string
+	PropertiesRequired []string
+	PropertiesOptional []string
+	PropertiesReadOnly []string
+	// The raw data from the meta file.
+	// Storing the raw data proactively in case we need to access the data at a later stage.
+	RawMetaData map[string]interface{}
```
Comment (lhercot) [inline gen/utils/data/class.go:56]: Is Raw if it is a map[string]interface{}?

File: gen/utils/data/class.go [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+	// Indicates if the class is required when defined as a child in a parent resource.
+	Required bool
```
Comment (lhercot) [inline gen/utils/data/class.go:58]: I think the name is not clear enough for this.

File: gen/utils/data/class.go [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+	// The resource name is the name of the resource in the provider, ex "aci_tenant".
+	ResourceName string
+	// The nested attribute name when part of a parent resource.
+	// ex. "aci_relation_from_bridge_domain_to_netflow_monitor_policy" would translate to "aci_relation_to_netflow_monitor_policy".
```
Comment (lhercot) [inline gen/utils/data/class.go:62]:
```suggestion
	// ex. "aci_relation_from_bridge_domain_to_netflow_monitor_policy" would translate to "relation_to_netflow_monitor_policy".
```

File: gen/utils/data/class.go [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+	ResourceNameNested string
+	// The relative name (RN) format of the class, ex "tn-{name}".
+	RnFormat string
+	// The supported APIC versions for the class.
+	// Each version range is separated by a comma, ex "4.2(7f)-4.2(7w),5.2(1g)-".
+	// The first version is the minimum version and the second version is the maximum version.
+	// A dash at the end of a range (ex. 4.2(7f)-) indicates that the class is supported from the first version to the latest version.
+	Versions string
```
Comment (lhercot) [inline gen/utils/data/class.go:70]: I would add DeprecatedVersions

Comment (lhercot) [inline gen/utils/data/class.go:70]: string or convert it already into list of ranges?

File: gen/utils/data/class.go [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+type Child struct {
+	// The name of the child class, ex "fvTenant".
+	// This is used as the key for the map of all classes.
+	ClassName string
+	// When it is a relationship class, this is the class to which the relationship points.
+	PointsToClass string
+}
+
+type ClassDocumentation struct {
+	Description string
```
Comment (lhercot) [inline gen/utils/data/class.go:82]: No comment for description?

File: gen/utils/data/class.go [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+		if unicode.IsUpper(character) {
+			shortName = className[index:]
+			packageName = className[:index]
+			break
+		}
+	}
+	if shortName == "" || packageName == "" {
+		genLogger.Fatal(fmt.Sprintf("Failed to split class name '%s' for name space separation.", className))
+	}
+
```
Comment (lhercot) [inline gen/utils/data/class.go:111]: Should we log the package and shortName here in Trace?

File: gen/utils/data/class.go [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+		ClassName:             className,
+		ClassNameShort:        shortName,
+		ClassNameForFunctions: cases.Title(language.Und, cases.NoLower).String(className),
+		ClassNamePackage:      packageName,
+		Properties:            make(map[string]*Property),
+	}
+}
+
+func (c *Class) LoadMetaFile() {
+	genLogger.Trace(fmt.Sprintf("Loading meta file for class '%s'.", c.ClassName))
```
Comment (lhercot) [inline gen/utils/data/class.go:122]:
```suggestion
	genLogger.Debug(fmt.Sprintf("Loading meta file for class '%s'.", c.ClassName))
```

File: gen/utils/data/class.go [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+
+	c.setResourceName()
+
+	// TODO: add functions to set the other class data
+
+	if properties, ok := c.RawMetaData["properties"]; ok {
+		c.setProperties(properties.(map[string]interface{}))
+	}
+
+	genLogger.Trace(fmt.Sprintf("Succesfully set class data for class '%s'.", c.ClassName))
```
Comment (lhercot) [inline gen/utils/data/class.go:154]:
```suggestion
	genLogger.Debug(fmt.Sprintf("Succesfully set class data for class '%s'.", c.ClassName))
```

File: gen/utils/data/property.go [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+	// Test specific information for the property.
+	// This is used to generate the test cases and exmaples for the property.
+	// TODO: re-evaluate the structure when creating example and test templates.
+	TestValues []TestValue
+	// Validation specific information for the property.
+	// In the meta file for the class this is a regex statement that is used to validate the property.
+	// TODO: re-evaluate the structure when creating resource templates. We might want to create a separate struc type for each type of validation.
+	Validators []Validator
+	// Specifies the valid values for the property when only certain values are allowed as input.
+	ValidValues []string
```
[the remark this answers is not in the document]
Comment (akinross) [reply gen/utils/data/property.go:39]: For now we do not, would need to investigate how we could add this functionality and raise enhancement to meta file generation to add versions to validValues. I will already create []ValidValue, where we specify the ValidValue as a type and read the available information.

File: gen/utils/data/property.go [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+	// The type of the legacy attribute.
+	Type string
+	// Indicates if a property is required in the legacy resource schema.
+	Required bool
+	// Indicates if a property is optional in the legacy resource schema.
+	Optional bool
+	// Indicates if a property is computed in the legacy resource schema.
+	Computed bool
+	// The name of the attribute in the legacy resource schema.
+	LegacyAttributeName string
```
[the remark this answers is not in the document]
Comment (akinross) [reply gen/utils/data/property.go:63]: The type is already specified in the type of MigrationValue. Not sure what the benefit would be for adding that information from schema/state perspective, release note documentation? For now think what we have is enough and we can adjust later when required. Looking at the type definition I will change the order and change LegacyAttributeName to Name, since it is contained.

File: gen/utils/data/property.go [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+	// - string: when the property is a single value
+	// - set: when the property is bitmask and alleows multiple values
+	// TODO: investigate enum type for this
+	ValueType string
+	// The supported APIC versions for the property.
+	// Each version range is separated by a comma, ex "4.2(7f)-4.2(7w),5.2(1g)-".
+	// The first version is the minimum version and the second version is the maximum version.
+	// A dash at the end of a range (ex. 4.2(7f)-) indicates that the class is supported from the first version to the latest version.
+	Versions string
+}
```
[the remark this answers is not in the document]
Comment (akinross) [reply gen/utils/data/property.go:51]: We can add when the use-case arises, don't think we did anything like this yet. Kept the types/structs to a minimum what I knew for now from scratch to be able to test that reading of class and property info is working. Could have limited it even more. Will enhance the class and property type, while I do all the functions for the types and when we work on the templates.

File: gen/utils/data/data.go [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+)
+
+const (
+	// Environment variables for logging configuration.
+	// Determine if the meta data should be refreshed from remote location, the default is to not refresh the meta data.
+	envMetaRefresh = "GEN_ACI_TF_META_REFRESH"
+	// The from which the meta data is retrieved, when not set, the default is the devnet base URL.
+	envMetaHost = "GEN_ACI_TF_META_HOST"
+	// The environment variable that contains the list of classes to be retrieved.
+	// The classes are separated by a comma, ex "fvTenant,fvAp".
```
[the remark this answers is not in the document]
Comment (akinross) [reply gen/utils/data/data.go:23]: then the classes are not retrieved, change the env var for refresh and added some more explanation

File: gen/utils/data/data.go [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+		// The creation of the class also sets the class name and package name which are required for the meta file download.
+		classDetails := *NewClass(className)
+		ds.Classes[className] = classDetails
+
+		if retrieve {
+			// ENHANCEMENT: Concurrently retrieve/write the meta data.
+			// Retrieve the meta file for the class from the remote location.
+			ds.retrieveMetaFileFromRemote(classDetails)
+		}
+	}
```
[the remark this answers is not in the document]
Comment (akinross) [reply gen/utils/data/data.go:114]: reason why I did it this way is to avoid looping the list of classes twice and to avoid retrieval twice, will move things around

File: gen/utils/data/class.go (hunk 4 above)
Comment (akinross) [reply gen/utils/data/class.go:56]: what do you mean?

File: gen/utils/data/class.go (hunk 10 above)
Comment (akinross) [reply gen/utils/data/class.go:122]: will change here, but there are more cases where I use trace, should these all be debug?

File: gen/utils/data/property.go [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+	// The following types are supported:
+	// - string: when the property is a single value
+	// - set: when the property is bitmask and alleows multiple values
+	// TODO: investigate enum type for this
+	ValueType string
+	// The supported APIC versions for the property.
+	// Each version range is separated by a comma, ex "4.2(7f)-4.2(7w),5.2(1g)-".
+	// The first version is the minimum version and the second version is the maximum version.
+	// A dash at the end of a range (ex. 4.2(7f)-) indicates that the class is supported from the first version to the latest version.
+	Versions string
```
[the remark this answers is not in the document]
Comment (akinross) [reply gen/utils/data/property.go]: changed

File: gen/utils/data/class.go (hunk 8 above)
Comment (akinross) [reply gen/utils/data/class.go:84]: added

File: gen/utils/data/class.go (hunk 9 above)
Comment (akinross) [reply gen/utils/data/class.go:126]: added

File: gen/utils/utils.go [hunk 18]
```diff
@@ -0,0 +1,14 @@
+package utils
+
+import "os"
+
+func GetFileNamesFromDirectory(path string) []string {
+	var names []string
+	entries, err := os.ReadDir(path)
```
[the remark this answers is not in the document]
Comment (akinross) [reply gen/utils/utils.go:16]: Did not do originally because I thought from generator point of view we control what files to add to this meta folder ( which should only be meta files ), and it would fail later on the file name. Have added directory check and test case for the function.

[the remark this answers is not in the document]
Comment (akinross) [reply gen/utils/utils.go:18]: Add additional functionality to the function to support removal of extension and added test cases for both directory and extension removal.

File: gen/utils/data/data.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (akinross) [reply gen/utils/data/data.go]: Restructured the code to de-couple

File: gen/utils/data/class.go [hunk 19]
```diff
@@ -0,0 +1,202 @@
...
+	for index, character := range className {
+		if unicode.IsUpper(character) {
+			shortName = className[index:]
+			packageName = className[:index]
+			break
+		}
+	}
+
+	if packageName == "" || shortName == "" {
+		genLogger.Trace(fmt.Sprintf("Class name '%s' got split into package name '%s' and short name '%s'.", className, packageName, shortName))
```
Comment (shrsr) [inline gen/utils/data/class.go:196]: Isn't this supposed to be outside the if statement at the bottom?

File: gen/utils/data/data.go [hunk 20]
```diff
@@ -0,0 +1,144 @@
...
+	retrievedClasses []string
+}
+
+func NewDataStore() *DataStore {
+	dataStore := &DataStore{Classes: make(map[string]Class)}
+	// Set the meta data host for retrieval of meta files.
+	dataStore.setMetaHost()
+	// Check if classes are set in the environment variable 'GEN_ACI_TF_META_CLASSES' and retrieve the meta files for those classes.
+	dataStore.retrieveEnvMetaClassesFromRemote()
+	// Refresh the meta files from the remote location if specified in the environment variable 'GEN_ACI_TF_META_REFRESH'.
```
Comment (shrsr) [inline gen/utils/data/data.go:36]: Probably need to add this here:
```
func NewDataStore(className){
dataStore.retrieveMetaFileFromRemote(className)
```

**Review by shrsr — CHANGES_REQUESTED** [review]

File: gen/utils/data/class.go (hunk 19 above)
Comment (akinross) [reply gen/utils/data/class.go:196]: was intended to display only when it failed but then would suggest to move above the conditional to always print the line, since the fail case is fatal

File: gen/utils/data/data.go (hunk 20 above)
Comment (akinross) [reply gen/utils/data/data.go:36]: Can you elaborate what you mean, because I do not understand your comment

File: gen/utils/data/property.go [hunk 21]
```diff
@@ -0,0 +1,130 @@
...
+	// Test specific information for the property.
+	// This is used to generate the test cases and examples for the property.
+	// TODO: re-evaluate the structure when creating example and test templates.
+	TestValues []TestValue
+	// Validation specific information for the property.
+	// In the meta file for the class this is a regex statement that is used to validate the property.
+	// TODO: re-evaluate the structure when creating resource templates. We might want to create a separate struct type for each type of validation.
+	Validators []Validator
+	// Specifies the valid values for the property when only certain values are allowed as input.
+	ValidValues []ValidValue
```
Comment (samiib) [inline gen/utils/data/property.go:42]: Have we thought about how to handle the case of a valid value being removed in the meta? Its potentially a metadata problem, but we do currently have an outstanding issue in the metadata update [PR](https://github.com/CiscoDevNet/terraform-provider-aci/pull/1314/files#diff-50786653bb423fd39c0e906cd63c60f32d159bd9627395ae17b4599da93b9448R261).

Comment (akinross) [reply gen/utils/data/property.go:42]: Good catch. Yes I have considered it (saw you created an issue for it on Jira) but currently do not see a way in meta where this is done in a versioned approach, so would not know yet how we should deal with this situation. We currently have overwrite mechanism where we could add valid/remove valid values so we can bypass the situation temporarily until we find an solution that is suitable from meta point and generator point of view. 

Furthermore since we are in phase 1 of redesign and we are not allowed to make any changes to the generated files (resource/datasource/docs/examples/tests) it would be hard to already work on a solution. I would suggest to handle this in phase 2 with the already opened Jira ticket.

I would think this evaluation should be done on a case by case basis ( where the change in behaviour comes to light because of the generator ) where we would need to evaluate if the value was even valid in the first case. In this case it seems that this `keyType` property was exposed with wrong valid values. Where cleanup activities of meta have removed it, but we would need to validate this with the meta files owner. And perhaps raise an enhancement request in CDETS process.

Depending on the outcome of that I think we should investigate a way to implement a solution. We could look into custom input validation with some form of warning message during a deprecation period, or keeping the valid value as input and letting APIC handle the error while making a note in the documentation.

File: gen/utils/data/data.go (hunk 20 above)
Comment (shrsr) [reply gen/utils/data/data.go:36]: You can ignore this, I thought we needed to call the method inside here

**Review by anvitha-jain — APPROVED** [review]
LGTM

File: gen/utils/data/property.go (hunk 21 above)
Comment (samiib) [reply gen/utils/data/property.go:42]: I agree, having the override mechanism should be enough for us to handle the case for now. Its an interesting case to handle. I was thinking the most straight forward way is to keep the value for backwards compatibility, add a note in the documentation and if possible log a warning when the value is being used. I'll link these comments to the backlog ticket.

File: gen/utils/data/class.go (hunk 19 above)
Comment (shrsr) [reply gen/utils/data/class.go]: We should probably move it after the if condition and before the return statement?

File: gen/utils/data/class.go [hunk 22]
```diff
@@ -0,0 +1,215 @@
...
+	genLogger.Trace(fmt.Sprintf("Splitting class name '%s' for name space separation.", className))
+	for index, character := range className {
+		if unicode.IsUpper(character) {
+			shortName = className[index:]
+			packageName = className[:index]
+			break
+		}
+	}
+
+	genLogger.Debug(fmt.Sprintf("Class name '%s' got split into package name '%s' and short name '%s'.", className, packageName, shortName))
```
Comment (shrsr) [inline gen/utils/data/class.go:207]: We should probably move this after the if packageName == "" || shortName == "" condition?

**Review by shrsr — CHANGES_REQUESTED** [review]

File: gen/utils/data/class.go (hunk 22 above)
Comment (akinross) [reply gen/utils/data/class.go:207]: Put it in front on purpose so that the split is always in the logs, even when there is a fatal condition. Do you prefer to have it after and change the logging of fatal to display what it got split into?

**Review by sajagana — CHANGES_REQUESTED** [review]

**Review by shrsr — APPROVED** [review]
LGTM

File: gen/generator.go [hunk 23]
```diff
@@ -1,14 +1,19 @@
...
 
 func main() {
+	genLogger.Trace("Starting Generator")
 	// Ensure that the log file is closed at the end of the main().
 	defer genLogger.CloseLogFile()
-	genLogger.Info("Implement Generator")
+
+	genLogger.Trace("Initializing Data Store")
+	data.NewDataStore()
+	genLogger.Trace("Initializing Data Store complete")
```
Comment (lhercot) [inline gen/generator.go:18]:
```suggestion
	genLogger.Trace("Initialization of Data Store complete")
```

File: gen/utils/data/class_test.go [hunk 24]
```diff
@@ -0,0 +1,28 @@
...
+	assert.Equal(t, packageName, "fv", fmt.Sprintf("Expected package name to be 'fv', but got '%s'", packageName))
+	assert.Equal(t, shortName, "Tenant", fmt.Sprintf("Expected short name to be 'Tenant', but got '%s'", shortName))
+}
+
+func TestSplitClassNameToPackageNameAndShortNameMultiple(t *testing.T) {
+	test.InitializeTest(t, 0)
+	packageName, shortName := splitClassNameToPackageNameAndShortName(constTestClassNameMultipleWordsInShortName)
+	assert.Equal(t, packageName, "fv", fmt.Sprintf("Expected package name to be 'fv', but got '%s'", packageName))
+	assert.Equal(t, shortName, "RsIpslaMonPol", fmt.Sprintf("Expected short name to be 'RsIpslaMonPol', but got '%s'", shortName))
+}
```
Comment (lhercot) [inline gen/utils/data/class_test.go:28]: Do we want a negative test that triggers the error "failed to split"?

File: gen/utils/data/class.go [hunk 25]
```diff
@@ -0,0 +1,215 @@
...
+
+type PlatformTypeEnum int
+
+// The enumeration options of the Platform type.
+const (
+	// Apic indicates that the class is available on the on-premises version of APIC.
+	Apic PlatformTypeEnum = iota + 1
+	// Cloud indicates that the class is available on the cloud version of APIC.
+	Both
+	// Both indicates that the class is available on both the on-premises and cloud versions of APIC.
```
Comment (lhercot) [inline gen/utils/data/class.go:82]: Flip comments.

File: gen/utils/data/data.go [hunk 26]
```diff
@@ -0,0 +1,144 @@
...
+
+	}
+}
+
+func (ds *DataStore) retrieveMetaFileFromRemote(className string) {
+	packageName, shortName := splitClassNameToPackageNameAndShortName(className)
+	url := fmt.Sprintf(constMetaFileUrl, ds.metaHost, packageName, shortName)
+	genLogger.Debug(fmt.Sprintf("Retrieving meta data for class '%s' from: %s.", className, url))
+
+	client := &http.Client{Transport: &http.Transport{TLSClientConfig: &tls.Config{InsecureSkipVerify: true}}}
```
Comment (lhercot) [inline gen/utils/data/data.go:109]: Do we need to initialize a new client for every API call? Shouldn't we store it in the DataStore itself?

File: gen/utils/data/data.go [hunk 27]
```diff
@@ -0,0 +1,144 @@
...
+	res, err := client.Get(url)
+	if err != nil {
+		genLogger.Fatal(fmt.Sprintf("Error during retrieval of meta file for class '%s': %s.", className, err.Error()))
+	}
+
+	resBody, err := io.ReadAll(res.Body)
```
Comment (lhercot) [inline gen/utils/data/data.go:115]: Do we need to read the whole file in memory?
Should we use io.Copy?

File: gen/utils/data/property.go (hunk 21 above)
Comment (lhercot) [reply gen/utils/data/property.go:42]: Should we add a []Versions in the ValidValue struct?

**Review by lhercot — CHANGES_REQUESTED** [review]

File: gen/utils/data/class_test.go (hunk 24 above)
Comment (akinross) [reply gen/utils/data/class_test.go:28]: I did not add tests for the negative flows that result into a fatal. I can change this by having the function returning errors, but decided not to because of simplicity. Let's discuss 1on1

File: gen/generator.go (hunk 23 above)
Comment (akinross) [reply gen/generator.go]: ok

File: gen/utils/data/class.go (hunk 25 above)
Comment (akinross) [reply gen/utils/data/class.go]: ok

File: gen/utils/data/data.go (hunk 26 above)
Comment (akinross) [reply gen/utils/data/data.go:109]: will change

File: gen/utils/data/property.go (hunk 21 above)
Comment (akinross) [reply gen/utils/data/property.go:42]: Added to valid value and also changed the defaults struct in documentation section

File: gen/utils/data/class_test.go (hunk 24 above)
Comment (akinross) [reply gen/utils/data/class_test.go:39]: changed by returning errors for functions

**Review by gmicol — APPROVED** [review]
LGTM

File: gen/utils/data/data.go [hunk 28]
```diff
@@ -0,0 +1,181 @@
...
+	"github.com/CiscoDevNet/terraform-provider-aci/v2/gen/utils"
+	"github.com/CiscoDevNet/terraform-provider-aci/v2/gen/utils/logger"
+)
+
+// Initialize a logger instance for the generator.
+var genLogger = logger.InitializeLogger()
+
+type DataStore struct {
+	// A map containing all the information about the classes required to render the templates.
+	Classes map[string]Class
```
Comment (lhercot) [inline gen/utils/data/data.go:23]: Does it need to be public?
```suggestion
	classes map[string]Class
```

File: gen/utils/data/data.go [hunk 29]
```diff
@@ -0,0 +1,181 @@
...
+	}
+
+	outputFile, err := os.Create(fmt.Sprintf("%s/%s.json", constMetaPath, className))
+	if err != nil {
+		genLogger.Error(fmt.Sprintf("Error during creation of file for class '%s': %s.", className, err.Error()))
+		return err
+	}
+
+	defer outputFile.Close()
+	_, err = outputFile.Write(buffer.Bytes())
```
Comment (lhercot) [inline gen/utils/data/data.go:155]: I thought we could do something like:
```suggestion
	outputFile, err := os.Create(fmt.Sprintf("%s/%s.json", constMetaPath, className))
	if err != nil {
		genLogger.Error(fmt.Sprintf("Error during creation of file for class '%s': %s.", className, err.Error()))
		return err
	}

	defer outputFile.Close()
	_, err = io.Copy(outputFile, res.Body)
```
Without the need to buffer it in memory.

File: gen/utils/data/data.go [hunk 30]
```diff
@@ -0,0 +1,181 @@
...
+	if err != nil {
+		genLogger.Error(fmt.Sprintf("Error during writing to file for class '%s': %s.", className, err.Error()))
+		return err
+	}
+
+	ds.retrievedClasses = append(ds.retrievedClasses, className)
```
Comment (lhercot) [inline gen/utils/data/data.go:161]: Any risk of duplication in this list?

**Review by lhercot — CHANGES_REQUESTED** [review]

File: gen/utils/data/data.go (hunk 28 above)
Comment (akinross) [reply gen/utils/data/data.go:23]: for now made it public because outside the data the classes should be accessed, can change and create method to loop through and retrieve information from the map if you think that would be better.

File: gen/utils/data/data.go (hunk 29 above)
Comment (akinross) [reply gen/utils/data/data.go:155]: good one, interesting!

File: gen/utils/data/data.go (hunk 30 above)
Comment (akinross) [reply gen/utils/data/data.go:161]: in theory it could, so moved the conditional inside the retrieveMetaFileFromRemote function

**Review by lhercot — APPROVED** [review]
LGTM
